File: examples/demo5.py
'''
Upper level demo that illustrates the computation of coil sensitivity maps,
applying projection from the image space into acquisition space and back
defined by the aquisition model, and images and acquisitions algebra.
'''

import math
import os
import pylab
import sys
import time

BUILD_PATH = os.environ.get('BUILD_PATH') + '/xGadgetron'
SRC_PATH = os.environ.get('SRC_PATH') + '/xGadgetron/pGadgetron'

sys.path.append(BUILD_PATH)
sys.path.append(SRC_PATH)

from pGadgetron import *

def main():
    # acquisitions will be read from this HDF file
    file = str(input('raw data file (with apostrophys in Python2.*): '))
    input_data = MR_Acquisitions(file)

    print('---\n acquisition data norm: %e' % input_data.norm())

    processed_data = MR_remove_x_oversampling(input_data)

    print('---\n processed acquisition data norm: %e' % processed_data.norm())

    # perform reconstruction
    recon = MR_BasicReconstruction()
    recon.set_input(processed_data)
    recon.process()
    complex_images = recon.get_output()

    print('---\n reconstructed images norm: %e' % complex_images.norm())

    csms = MR_CoilSensitivityMaps()

##    csm_file = str(input('csm file: '))
##    print('reading sensitivity maps...')
##    csms.read(csm_file)

    print('---\n sorting acquisitions...')
    processed_data.sort()
    print('---\n computing sensitivity maps...')
    csms.calculate(processed_data)

    # create acquisition model based on the acquisition parameters
    # stored in input_data and image parameters stored in complex_images
    am = MR_AcquisitionModel(processed_data, complex_images)

    am.set_coil_sensitivity_maps(csms)

    # use the acquisition model (forward projection) to produce 'acquisitions'
    acqs = am.forward(complex_images)

    print('---\n reconstructed images forward projection norm %e' % acqs.norm())

    # compute the difference between real and modelled acquisitions
    diff = acqs - processed_data
    rr = diff.norm()/acqs.norm()
    print('---\n reconstruction residual norm (rel): %e' % rr)

    # compare reconstructed images with backward-projected acquisitions
    bwd_images = am.backward(processed_data)
    im_diff = bwd_images - complex_images
    print(\
        '---\n difference between reconstructed and back-projected images: %e'\
        % (im_diff.norm()/complex_images.norm()))

    # test that the backward projection is the adjoint of forward
    # on x = processed_data and y = complex_images
    xFy = processed_data * acqs
    Bxy = bwd_images * complex_images
    print('---\n (x, F y) = (%e, %e)' % (xFy.real, xFy.imag))
    print('= (B x, y) = (%e, %e)' % (Bxy.real, Bxy.imag))

    # extract real images from complex
    images = MR_extract_real_images(complex_images)

    # plot obtained images
    for i in range(images.number()):
        data = images.image_as_array(i)
        pylab.figure(i + 1)
        pylab.imshow(data[0,0,:,:])
        print('Close Figure %d window to continue...' % (i + 1))
        pylab.show()

try:
    main()
except error as err:
    # display error information
    print ('Gadgetron exception occured:\n', err.value)

File: examples/demo5.m
% Upper-level demo, illustrates the use of acquisition models and images and acquisitions algebra.
% Involves the computation of coil sensitivity maps and
% the projection from the image space into acquisition space and back.

if ~libisloaded('mgadgetron')
    loadlibrary('mgadgetron')
end
if ~libisloaded('mutilities')
    loadlibrary('mutilities')
end

try
    % acquisitions will be read from this HDF file
    input_data = gadgetron.MR_Acquisitions('testdata.h5');
    fprintf('%d acquisitions found\n', input_data.number())
    
    % pre-process acquisition data
    fprintf('processing acquisitions...\n')
    processed_data = gadgetron.MR_remove_x_oversampling(input_data);
	
    % perform reconstruction
    recon = gadgetron.MR_BasicReconstruction();
    recon.set_input(processed_data)
    fprintf('reconstructing...\n')
    recon.process()
    complex_images = recon.get_output();
    
    % post-process reconstructed images
    fprintf('processing images...\n')
    images = gadgetron.MR_extract_real_images(complex_images);

    csms = gadgetron.MR_CoilSensitivityMaps();
    fprintf('sorting acquisitions...\n')
    processed_data.sort()
    fprintf('calculating sensitivity maps...\n')
    csms.calculate(processed_data)

    % create acquisition model based on the acquisition parameters
    % stored in input_data and image parameters stored in interim_images
    am = gadgetron.MR_AcquisitionModel(processed_data, complex_images);

    am.set_coil_sensitivity_maps(csms)

    % use the acquisition model (forward projection) to produce acquisitions
    acqs = am.forward(complex_images);

    % compute the difference between real and modelled acquisitions
    diff = acqs - processed_data;
    fprintf('reconstruction residual: %e\n', diff.norm()/acqs.norm())

    % apply the adjoint model (backward projection)
    imgs = am.backward(processed_data);

    % test that the backward projection is the adjoint of forward
    % on x = diff and y = complex_images
    fprintf('(x, F y) = %s\n', num2str(processed_data * acqs))
    fprintf('= (B x, y) = %s\n', num2str(imgs * complex_images))

    % test images norm and dot product
    s = imgs.norm();
    fprintf('(B x, B x) = %s = %e\n', num2str(imgs * imgs), s*s)

    % test linear combination of images
    im_diff = imgs - complex_images;
    fprintf('0.0 = %e\n', im_diff.norm()/complex_images.norm())

    % plot obtained images
    for i = 1 : images.number()
        data = images.image_as_array(i);
        figure(1000000 + i)
        data = data/max(max(max(data)));
        imshow(data(:,:,1,1));
    end

catch err
    % display error information
    fprintf('%s\n', err.message)
    fprintf('error id is %s\n', err.identifier)
end
